server/src/main/java/org/elasticsearch/common/util/concurrent/EsAbortPolicy.java

```diff
@@ -8,26 +8,27 @@
 
 package org.elasticsearch.common.util.concurrent;
 
-import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.ThreadPoolExecutor;
 
 public class EsAbortPolicy extends EsRejectedExecutionHandler {
 
     @Override
     public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
-        if (r instanceof AbstractRunnable) {
-            if (((AbstractRunnable) r).isForceExecution()) {
-                BlockingQueue<Runnable> queue = executor.getQueue();
-                if ((queue instanceof SizeBlockingQueue) == false) {
-                    throw new IllegalStateException("forced execution, but expected a size queue");
+        if (r instanceof AbstractRunnable abstractRunnable) {
+            if (abstractRunnable.isForceExecution()) {
+                if (executor.getQueue() instanceof SizeBlockingQueue<Runnable> sizeBlockingQueue) {
+                    try {
+                        sizeBlockingQueue.forcePut(r);
+                        return;
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                        throw new IllegalStateException("forced execution, but got interrupted", e);
+                    }
                 }
-                try {
-                    ((SizeBlockingQueue<Runnable>) queue).forcePut(r);
-                } catch (InterruptedException e) {
-                    Thread.currentThread().interrupt();
-                    throw new IllegalStateException("forced execution, but got interrupted", e);
+
+                if (executor.isShutdown() == false) {
```

Comment:
Another possible fix here is to use a different rejection handler if the executor has an unbounded queue.

Reply:
If the executor is shutting down, then this condition evaluates to false, and we fall through to reject execution below, right? Yet we may have already forced the task onto the queue, so don't we need to try to take it off before claiming it is rejected, like in EsExecutors.ForceQueuePolicy? Actually why do we even have two rejected execution handlers at all?

Reply:
If we forced the task onto the queue then we don't get here, we already called `return;`. But yes, it looks like there's a bug here too, see https://github.com/elastic/elasticsearch/issues/104580#issuecomment-1900462181.

> Actually why do we even have two rejected execution handlers at all?

Why indeed.

Reply:
AFAICT the different handlers correspond with the different types of queue in use. Therefore it seems we should have another handler for the unbounded `LinkedTransferQueue` that causes this failure - see 212e813dc8d7df7557b6f30a1bf66409440a72f5.